### PR TITLE
p5: send custom events in order to control the behavior of the draw loop

### LINKED ISSUE #34: p5: send custom events in order to control the behavior of the draw loop
Currently the is no way to control the execution of Proc draw loop since it is handled by events received by a receive only channel (<-chan event.Event) https://github.com/go-p5/p5/blob/fea8ed50309696b61da315102400b5dec754a593/proc.go#L208). 

Providing a way to send events and control/stop the loop would make Proc more testable.

### PR DESCRIPTION
**Pull request by psampaz**
closes #34

### CODE REVIEW

File: api_test.go [hunk 1]
```diff
@@ -327,3 +330,26 @@ func TestRandomSeed(t *testing.T) {
 		t.Errorf("Calling RandomSeed with different seeds should produce different sequence of numbers")
 	}
 }
+
+func TestProc(t *testing.T) {
```
Comment (psampaz) [inline api_test.go:334]: this is just an example to demonstrate how someone can start and stop the Proc.run inside a test

**Review by sbinet — COMMENTED** [review]
nice.

`testProc` could be adapted to re-use more of that new infrastructure.

another possibility could be to be able to modify the `app.NewWindow` call with, say, `p.newWindow(...)`.
where `newWindow` is a field of `Proc`.

in normal usage, this would of course return `app.NewWindow(...)`.
in tests, this would return a value that has an internal channel that can be controled from `testProc`.

e.g.:
```go
type Proc struct {
    ...
    newWindow func(opts ...app.Option) events
}

func newProc(...) *Proc {
    ...
    p.newWindow = func(opts ...app.Options) events {
        return app.NewWindow(opts...)
    }
}

type events interface {
    Events() <-event.Event
}
```

with this alternate design, one can reuse the whole `Proc.run` method w/o introducing a `select`.

WDYT?

**Comment by sbinet** [comment]
well, "of course" the `events` interface needs to be somewhat expanded to, say:
```go
type gioWindow interface {
    Events() <-chan event.Event
    Invalidate()
    Close()
}
```

**Comment by psampaz** [comment]
using the interface is like a much cleaner way to separate tests from normal code. I will give it a try.

**Comment by psampaz** [comment]
@sbinet please have a look

https://github.com/go-p5/p5/pull/35/commits/8e0e9d3251929209a0ef310f1e82828cb3ca9ae4 : Solution using Proc event field
https://github.com/go-p5/p5/pull/35/commits/9b4dda8223424482b0810eeb70ac38a63087602d : Solution using the interface

the data race in the failing test is related with the demo test here https://github.com/go-p5/p5/pull/35/files#diff-9a6ab27c824fb3a6082d5d518db17f3c310f1eff5def98ca5f5158309e92a08bR334. I didn't have the time to look into it yet.

File: proc.go [hunk 2]
```diff
@@ -54,6 +55,21 @@ var (
...
+	Events() <-chan event.Event
+	// Invalidates the window such that a FrameEvent will be generated immediately.
+	Invalidate()
+	// Close the window. The window's event loop should exit when it receives
+	// system.DestroyEvent.
+	Close()
+}
+
+// gioAppWindowFunc returns a function that creates Gio App.Window
+var gioAppWindowFunc = func(opts ...app.Option) appWindow {
```
Comment (sbinet) [inline proc.go:69]:
```suggestion
func gioAppWindow(opts ...app.Option) appWindow {
```

File: proc.go [hunk 3]
```diff
@@ -84,13 +100,15 @@ type Proc struct {
...
-	ctx  layout.Context
-	stk  *stackOps
-	head *headless.Window
-	rand *rand.Rand
+	ctx       layout.Context
+	stk       *stackOps
+	head      *headless.Window
+	rand      *rand.Rand
+	newWindow func(opts ...app.Option) appWindow
+	window    appWindow
```
Comment (sbinet) [inline proc.go:108]: I don't think we really need that new `window` field, do we?

Comment (sbinet) [inline proc.go:108]: I don't think we really need that new `window` field, do we?

File: proc.go [hunk 4]
```diff
@@ -84,13 +100,15 @@ type Proc struct {
...
 }
 
-func newProc(w, h int) *Proc {
+func newProc(w, h int, appWindowFunc func(opts ...app.Option) appWindow) *Proc {
```
Comment (sbinet) [inline proc.go:111]: let's drop that extra parameter.

File: proc.go [hunk 5]
```diff
@@ -108,6 +126,8 @@ func newProc(w, h int) *Proc {
 	proc.cfg.th = material.NewTheme(gofont.Collection())
 	proc.stk.cur().stroke.style.Width = 2
 
+	proc.newWindow = appWindowFunc
```
Comment (sbinet) [inline proc.go:129]:
```suggestion
	proc.newWindow = appWindow
```
just default to the "normal" `func`.

File: proc.go [hunk 6]
```diff
@@ -181,10 +201,11 @@ func (p *Proc) run() error {
 		height = p.cfg.h
 	)
 
-	w := app.NewWindow(app.Title("p5"), app.Size(
+	p.window = p.newWindow(app.Title("p5"), app.Size(
```
Comment (sbinet) [inline proc.go:204]:
```suggestion
	w := p.newWindow(app.Title("p5"), app.Size(
```

File: proc_test.go [hunk 7]
```diff
@@ -14,14 +14,49 @@ import (
...
 	"gioui.org/op"
 	"github.com/go-p5/p5/internal/cmpimg"
 )
 
 var GenerateTestData = flag.Bool("regen", false, "Uses the current state to regenerate the test data.")
 
+// testAppWindowFunc returns a function that creates Test App.Window
+var testAppWindowFunc = func(opts ...app.Option) appWindow {
+	return newTestGioWindow(opts...)
+}
```
Comment (sbinet) [inline proc_test.go:30]:
```suggestion
func testAppWindow(evts chan event.Event) func(opts ...app.Option) appWindow {
    return func(opts ...app.Option) appWindow {
        return testAppWindow{out: evts}
    }
}
```

File: proc_test.go [hunk 8]
```diff
@@ -38,7 +73,7 @@ func newTestGProc(t *testing.T, w, h int, setup, draw func(p *Proc), fname strin
 func newTestProc(t *testing.T, w, h int, setup, draw func(p *Proc), fname string) *testProc {
 	t.Helper()
 
-	p := newProc(w, h)
+	p := newProc(w, h, testAppWindowFunc)
```
Comment (sbinet) [inline proc_test.go:76]:
```suggestion
	p := newProc(w, h)
	ctl := make(chan event.Event)
	p.newWindow = testAppWindow(ctl)
```

and a new field to `testProc`:
```go
type testProc struct {
    ...
    ctl chan event.Event
}
```
which will be initialized with that `ctl` variable above.

File: proc_test.go [hunk 9]
```diff
@@ -14,14 +14,49 @@ import (
...
+// testAppWindowFunc returns a function that creates Test App.Window
+var testAppWindowFunc = func(opts ...app.Option) appWindow {
+	return newTestGioWindow(opts...)
+}
+
+// testAppWindow implements the appWindow interface and its purpose is to be used
+// in tests
+type testAppWindow struct {
+	out     chan event.Event
+	options []app.Option
```
Comment (sbinet) [inline proc_test.go:36]:
```suggestion
	evts chan event.Event
	opts []app.Option
```

File: proc_test.go [hunk 10]
```diff
@@ -14,14 +14,49 @@ import (
...
+
+func (t *testAppWindow) Invalidate() {
+	t.out <- system.FrameEvent{
+		Frame: func(ops *op.Ops) {},
+	}
+}
+
+func (t *testAppWindow) Close() {
+	t.out <- system.DestroyEvent{}
+}
```
Comment (sbinet) [inline proc_test.go:58]: I don't think there's a need to implement the interface as a pointer-receiver, is there?

**Review by sbinet — COMMENTED** [review]
thanks for this.

I may have an alternative design, see below.
WDYT?

File: proc.go (hunk 2 above)
Comment (sbinet) [reply proc.go:69]: or `func newAppWindow`

**Comment by sbinet** [comment]
if you prefer, I can send a PR with "my" alternate design (that was heavily inspired from your interesting out-of-the-box thinking) and we can compare notes.

as you prefer.

**Comment by psampaz** [comment]
Yes please do so, so  we can avoid the back and forth discussions. 

I didn't actually spend too much time on this, I just wanted to see if my understanding was close to what you suggested. 

btw I used the extra window field in order to be able to access the out event channel directly, since  the Event() method of the interface has to return a receive only channel (https://github.com/gioui/gio/blob/main/app/window.go#L112)

**Comment by psampaz** [comment]
closing this in favour of #37

**Pull request closed without merging**